### PR TITLE
Add Lago SDK

### PR DESCRIPTION
**Pull request by nvgoldin**
This PR adds a new SDK capability to Lago, the reasoning behind it is:
1. Start/destroy/initialize the environment directly from Python. Without using the ``in_lago_prefix`` flow.
2. "Formulate" the SDK we already provide(which is used by OST), and have more fine-grained control on what we expose, and what should be kept internal for the CLI.

A quick [example](https://gist.github.com/nvgoldin/f89b2bac5c61a2f352dbd5e1734e25a4) of what can be done with the SDK and PyTest now(the init file included for clarity):

```python
import pytest
import textwrap
import tempfile

from lago import sdk


def generate_config():
    conf = """
    domains:
      vm-el73:
        nics:
          - net: lago
        disks:
          - template_name: el7.3-base
            type: template
            name: root
            dev: sda
            format: qcow2
    nets:
      lago:
        type: nat
        dhcp:
          start: 100
          end: 254
        management: true
        dns_domain_name: lago.local
    """
    with tempfile.NamedTemporaryFile(delete=False) as f:
        f.write(textwrap.dedent(conf))

    return f.name


@pytest.fixture(scope='module')
def env():
    env = sdk.init(config=generate_config(), workdir='/tmp/pytest-lago')
    env.start()
    yield env
    env.destroy()


@pytest.fixture(scope='module')
def vm(env):
    vm = env.get_vms()['vm-el73']
    return vm


def test_ls(vm):
    res = vm.ssh(['ls', '-lsah'])
    assert res.code == 0


def test_hostname(vm):
    res = vm.ssh(['hostname -f'])
    assert res.code == 0
    assert str.strip(res.out) == 'vm-el73.lago.local'

```
and running pytest:
```bash
> pytest -vvv -s lago_sdk_example.py 
================================================================================================ test session starts ================================================================================================
platform linux2 -- Python 2.7.13, pytest-3.0.7, py-1.4.33, pluggy-0.4.0 -- /home/ngoldin/virtualenv/lago-venv/bin/python2
cachedir: .cache
rootdir: /home/ngoldin/src/nvgoldin.github.com/lago-init-files, inifile:
plugins: cov-2.4.0, hypothesis-3.8.2
collected 2 items 

lago_sdk_example.py::test_ls PASSED
lago_sdk_example.py::test_hostname PASSED

============================================================================================= 2 passed in 28.68 seconds =============================================================================================
```

### CODE REVIEW

File: lago/sdk.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    defaults['virt_config'] = config
+    workdir, prefix = cmd.do_init(**defaults)
+    return sdk(workdir, prefix)
+
+
+class sdk(object):
+    """
+    The SDK can be initialized in 2 ways:
+    1. (Preferred) - by calling :func:`sdk.init`.
+    2. By passing already created workdir and prefix objects.
```
Comment (gbenhaim) [inline lago/sdk.py]: Can we pass the path to the workdir instead? 
Maybe add a static method ```sdk.get_from_path(....)``` or something similar.

File: lago/sdk.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        'template_store': '/var/lib/lago/store',
+        'template_repos': '/var/lib/lago/repos',
+        'set_current': True,
+        'skip_bootstrap': False,
+        'skip_build': False,
+        'prefix_name': 'default',
+    }
+    defaults.update(kwargs)
+    defaults['virt_config'] = config
+    workdir, prefix = cmd.do_init(**defaults)
```
Comment (gbenhaim) [inline lago/sdk.py]: I think that ```lago.cmd``` should use ```lago.sdk```, and not the other way around.

File: lago/sdk.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    def __getattr__(self, name):
+        try:
+            attr = super(sdk, self).__getattr__(name)
+        except AttributeError:
+            attr = getattr(self._pprefix, name)
+        return attr
+
+    def __dir__(self):
+        return sorted(
```
Comment (gbenhaim) [inline lago/sdk.py]: Can you add docs with explanation?

File: lago/sdk.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        'prefix_name': 'default',
+    }
+    defaults.update(kwargs)
+    defaults['virt_config'] = config
+    workdir, prefix = cmd.do_init(**defaults)
+    return sdk(workdir, prefix)
+
+
+class sdk(object):
+    """
```
Comment (gbenhaim) [inline lago/sdk.py]: should be ```Sdk```  or ```SDK``` ?

File: lago/sdk_utils.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    Args:
+        attr(attribute): Attribute as returned by :func:`getattr`.
+        name(str): Attribute name.
+
+    Returns:
+        `attr` if passed.
+    """
+
+    if name.startswith('_'):
+        return attr
```
Comment (gbenhaim) [inline lago/sdk_utils.py]: Do we want to allow access to all private attributes?

File: lago/sdk_utils.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import wrapt
+
+
+class SDKWrapper(wrapt.ObjectProxy):
+    """A proxy object that exposes only methods which were decorated with
+    :func:`expose` decorator."""
+
+    def __getattr__(self, name):
+        attr = getattr(self.__wrapped__, name)
+        return filter_sdk(attr, name)
```
Comment (gbenhaim) [inline lago/sdk_utils.py]: I think that the name filter is misleading, I would go with ```check_if_exposed``` or ```validate_exposed``` or something similar.

File: lago.spec.in [hunk 7]
```diff
@@ -100,6 +101,7 @@ Requires: python-stevedore
 Requires: python-yaml
 Requires: python-enum
 Requires: pyxdg
+Requires: python-wrapt
```
Comment (gbenhaim) [inline lago.spec.in]: Is this package available on rhel ?

Comment (nvgoldin) [inline lago.spec.in]: EPEL.

File: lago/sdk.py (hunk 2 above)
Comment (nvgoldin) [inline lago/sdk.py]: I Agree, I think that will involve much more re factoring that I think is wanted for this patch. I wanted to provide the minimum framework needed in order to start using the SDK, and in parallel let us refactor it without breaking existing features.

File: lago/sdk.py (hunk 4 above)
Comment (nvgoldin) [inline lago/sdk.py]: Right, wasn't sure what is proper here.. our convention is 'Sdk', but it should be 'SDK', but that sounds like a constant :)
Lets go with SDK.

File: lago/sdk.py (hunk 1 above)
Comment (nvgoldin) [inline lago/sdk.py]: +1 - on my to do list - let initialize an environment from an existing path. Will formulate all these to-dos into an issue.

File: lago/sdk.py (hunk 3 above)
Comment (nvgoldin) [inline lago/sdk.py]: Will do.

File: lago/sdk_utils.py (hunk 5 above)
Comment (nvgoldin) [inline lago/sdk_utils.py]: Technically:
1. It gave me a few problems removing those(not actually the 
``_`` methods, but the magic methods(``__``), I guess we could pass the `__`. I'll have to check if `wrapt` handles it well. 
2. I think its OK keep the private methods. Having an underscore already indicates they are private(and most autocompleters would not complete them by default, well, at least my vim plugins).

File: lago/sdk_utils.py (hunk 6 above)
Comment (nvgoldin) [inline lago/sdk_utils.py]: Will change.

File: lago/sdk.py (hunk 1 above)
Comment (nvgoldin) [inline lago/sdk.py]: https://github.com/lago-project/lago/issues/532

File: lago/sdk.py (hunk 3 above)
Comment (nvgoldin) [inline lago/sdk.py]: Done.

File: lago/sdk_utils.py (hunk 6 above)
Comment (nvgoldin) [inline lago/sdk_utils.py]: Went with ``getattr_sdk`` - the reasoning is that it behaves similarly to ``getattr`` - raises ``AttributeError`` or returns the attribute.

**Comment by nvgoldin** [comment]
PR v2 + PR v3 - address @gbenhaim comments. Will squash eventually(but now we can see changes from patch to patch !:)

File: lago/sdk_utils.py (hunk 5 above)
Comment (gbenhaim) [inline lago/sdk_utils.py]: So what's the idea of using the proxy if we expose all the private attributes?

Comment (nvgoldin) [inline lago/sdk_utils.py]: I think that a "good" UX is when you import the SDK you only have public methods you are supposed to use and have documentation. If you "look" at the underscore methods, you know its on your own risk. You can also access ``SDK._prefix`` and get the "normal" object directly, but again that is internal and can change.

 As to why (still) use the proxy - we have many other "public" methods in prefix that are either:
1. Undocumented.
2. Do not make sense for SDK use(but do for CLI)(like ``save``, ``initialize``, ``resolve_parent``, ``fetch_url``, ``virt_conf_from_stream``, etc).

With the proxy - we can gradually change the internal implementation without breaking the SDK - like moving the 'do_init' to the SDK, etc.

I can filter the '_' methods also, its not a big issue.

Comment (nvgoldin) [inline lago/sdk_utils.py]: OK - removed the entire detection in https://github.com/lago-project/lago/pull/531/commits/4d184a839a7cce651ac9d1fb50e109a855f67850
Will only add "expose" methods.

**Comment by nvgoldin** [comment]
ci test please

**Comment by nvgoldin** [comment]
ci test please

**Comment by nvgoldin** [comment]
ci merge please

**Comment by nvgoldin** [comment]
ci merge please

**Pull request merged**